pkg/scheduler/util/non_zero.go

```diff
@@ -38,41 +39,92 @@ const (
 
 // GetNonzeroRequests returns the default cpu and memory resource request if none is found or
 // what is provided on the request.
-func GetNonzeroRequests(requests *v1.ResourceList) (int64, int64) {
-	return GetNonzeroRequestForResource(v1.ResourceCPU, requests),
-		GetNonzeroRequestForResource(v1.ResourceMemory, requests)
+func GetNonzeroRequests(pod *v1.Pod) (int64, int64) {
+	return GetNonzeroRequestForResource(pod, v1.ResourceCPU),
+		GetNonzeroRequestForResource(pod, v1.ResourceMemory)
 }
 
 // GetNonzeroRequestForResource returns the default resource request if none is found or
 // what is provided on the request.
-func GetNonzeroRequestForResource(resource v1.ResourceName, requests *v1.ResourceList) int64 {
-	switch resource {
-	case v1.ResourceCPU:
-		// Override if un-set, but not if explicitly set to zero
-		if _, found := (*requests)[v1.ResourceCPU]; !found {
-			return DefaultMilliCPURequest
+func GetNonzeroRequestForResource(pod *v1.Pod, resource v1.ResourceName) int64 {
+	var podRequest int64
+
+	if len(pod.Spec.Containers) == 0 {
+		return podRequest
+	}
+
+	if resource == v1.ResourceEphemeralStorage && !utilfeature.DefaultFeatureGate.Enabled(features.LocalStorageCapacityIsolation) {
+		// if the local storage capacity isolation feature gate is disabled, pods request 0 disk.
+		return podRequest
+	}
+
+	// The resource request may be assigned by the explicit 0. We need a state variable
+	// to indicate if the explicit resource request was found.
+	isExplicitRequest := false
+	for i := range pod.Spec.Containers {
+		container := &pod.Spec.Containers[i]
+		requests := container.Resources.Requests
+
+		switch resource {
+		case v1.ResourceCPU:
+			if request, found := requests[v1.ResourceCPU]; found {
+				podRequest += request.MilliValue()
+				isExplicitRequest = true
+			}
+		// memory, storage, ephemeral-storage, and other scaler extended resources.
+		default:
+			if request, found := requests[resource]; found {
+				podRequest += request.Value()
+				isExplicitRequest = true
+			}
```

Comment:
To not change the original logic, an you add another switch case for `v1.ResourceMemory` and `v1.ResourceEphemeralStorage` above. So that in this default case, we can additionally follow original logic to check `f v1helper.IsScalarResourceName(resource)`.

BTW: the whole switch block seems to be able to be re-used in the latter initContainers loop.

Reply:
+1

Reply:
have refactored